File: backend/app/schemas/supplier_unit_price.py
"""供应商工种单价 Schema。"""
from __future__ import annotations
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field
from app.schemas.base import FinanceDeskBaseModel


class SupplierUnitPriceCreate(FinanceDeskBaseModel):
    supplier_id: int = Field(..., description="供应商ID")
    year: int = Field(..., description="年度")
    laborer_price: Optional[float] = Field(None, description="普工单价")
    technician_price: Optional[float] = Field(None, description="技工单价")
    senior_technician_price: Optional[float] = Field(None, description="高级技工单价")
    special_work_price: Optional[float] = Field(None, description="特种作业单价")
    comprehensive_price: Optional[float] = Field(None, description="综合单价")
    remark: Optional[str] = Field(None, description="备注")


class SupplierUnitPriceUpdate(FinanceDeskBaseModel):
    year: Optional[int] = None
    laborer_price: Optional[float] = None
    technician_price: Optional[float] = None
    senior_technician_price: Optional[float] = None
    special_work_price: Optional[float] = None
    comprehensive_price: Optional[float] = None
    remark: Optional[str] = None


class SupplierUnitPriceResponse(FinanceDeskBaseModel):
    id: int
    supplier_id: int
    year: int
    laborer_price: Optional[float] = None
    technician_price: Optional[float] = None
    senior_technician_price: Optional[float] = None
    special_work_price: Optional[float] = None
    comprehensive_price: Optional[float] = None
    remark: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    is_deleted: bool


class PaginatedSupplierUnitPrices(FinanceDeskBaseModel):
    items: list[SupplierUnitPriceResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/services/supplierUnitPrice.ts
/* 🅵-1: supplierUnitPrice.ts - kill the `any` */
import { api } from './api';

export interface SupplierUnitPriceRecord {
  id: number;
  supplier_id: number;
  year: number;
  laborer_price: number | null;
  technician_price: number | null;
  senior_technician_price: number | null;
  special_work_price: number | null;
  comprehensive_price: number | null;
  remark: string | null;
  created_at: string;
  updated_at: string;
  is_deleted: boolean;
}

export interface SupplierUnitPriceCreatePayload {
  supplier_id: number;
  year: number;
  laborer_price?: number | null;
  technician_price?: number | null;
  senior_technician_price?: number | null;
  special_work_price?: number | null;
  comprehensive_price?: number | null;
  remark?: string | null;
}

export const fetchSupplierUnitPrices = (params?: { supplier_id?: number; year?: number; page?: number; page_size?: number }) => {
  const q = new URLSearchParams();
  q.set('page', String(params?.page || 1));
  q.set('page_size', String(params?.page_size || 20));
  if (params?.supplier_id) q.set('supplier_id', String(params.supplier_id));
  if (params?.year) q.set('year', String(params.year));
  return api.get<{ items: SupplierUnitPriceRecord[]; total: number; page: number; page_size: number }>('/supplier-unit-prices?' + q.toString());
};

export const createSupplierUnitPrice = (data: SupplierUnitPriceCreatePayload) =>
  api.post<SupplierUnitPriceRecord>('/supplier-unit-prices', data);

export const updateSupplierUnitPrice = (id: number, data: Partial<SupplierUnitPriceCreatePayload>) =>
  api.patch<SupplierUnitPriceRecord>(`/supplier-unit-prices/${id}`, data);

export const deleteSupplierUnitPrice = (id: number) =>
  api.delete<void>(`/supplier-unit-prices/${id}`);
